posts/inside-rust/2023-08-30-crates-io-malware-postmortem.md

```diff
@@ -0,0 +1,55 @@
+
+## Actions taken
+
+The crates were yanked and the associated user account was locked immediately after the crates.io team received the report. This would not have prevented someone from downloading the crate files directly from static.crates.io, but removed the crates from the crate index and made them uninstallable through normal `cargo` usage.
+
+After [analysing](#analysis) the crates in question and the logs related to all actions taken by this user — from signing up, searching for crates to typosquat, and finally to publishing the crates — we decided to delete the crates entirely from static.crates.io to prevent any further possibility of attack.[^deletion] The deletion was performed at 17:22 UTC on August 18, 2023.
+
+## Analysis
+
+In this case, the crates were solely and very obviously crafted for malicious purposes.
+
+We have no evidence that any of these crates were downloaded by an actual user — analysis of the user agents associated with the download requests for these crates in our logs suggest that the only downloads were automated scanner and mirroring actions.
+
+The user associated with these crates took no other actions — malicious or otherwise — in the 30 days[^retention] prior to August 18.
+
+## Future actions
+
+The [Rust Foundation][foundation]'s [Security Initiative][init] is planning future work on scanning all crate uploads, both in terms of typosquatting and the actual contents of the crate files. The prototypes of both projects would have detected these crate files, and the crates.io team intends to work with the Foundation to implement these scanners once ready.
+
+## Thanks
+
+Our thanks to Louis Lang at [Phylum][phylum] for reporting the crates, Josh Stone for facilitating the report, and Walter Pearce at the [Rust Foundation][foundation] for assisting with the analysis.
+
+[^deletion]: The crates were preserved for future analysis should there be other attacks, and to inform scanning efforts in the future.
+[^geolocation]: One wonders why they couldn't do this after sending the IP address to their Telegram channel, but federated workloads are apparently all the rage in 2023.
+[^putty]: [PuTTY][putty] is a popular SSH client for Windows.
+[^retention]: One year of logs are retained on crates.io, but only 30 days are immediately available on our log platform. We chose not to go further back in our analysis, since IP address based analysis is limited by the use of dynamic IP addresses in the wild, and the relevant IP address being part of an allocation to a residential ISP.
+[^telegram]: [Telegram][telegram] is a popular instant messaging app.
+[^typosquatting]: [Typosquatting][typosquatting] is a technique used by bad actors to initiate dependency confusion attacks where a legitimate user might be tricked into using a malicious dependency instead of their intended dependency — for example, a bad actor might try to publish a crate at `proc-macro` to catch users of the legitimate `proc-macro2` crate.
```

Comment:
```suggestion
[^typosquatting]: [Typosquatting][typosquatting] is a technique used by bad actors to initiate dependency confusion attacks where a legitimate user might be tricked into using a malicious dependency instead of their intended dependency — for example, a bad actor might try to publish a crate at `procmacro2` to catch users of the legitimate `proc-macro2` crate.
```

we should use a different example since iirc `proc-macro` is specifically prohibited since it comes with rustc

Reply:
I actually chose it _because_ it was blocked, just in case anyone got any bright ideas after reading the blog post. :laughing: